### PR TITLE
[fix][test] Fix ManagedCursorTest.testForceCursorRecovery

### LINKED ISSUE #23417: Flaky-test: ManagedCursorTest.testForceCursorRecovery
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/pulsar/issues) and found nothing similar.

### Example failure

https://github.com/apache/pulsar/actions/runs/11232853373/job/31225793977?pr=23414#step:11:2927

### Exception stacktrace

```
  Error:  Tests run: 282, Failures: 1, Errors: 0, Skipped: 202, Time elapsed: 36.683 s <<< FAILURE! - in org.apache.bookkeeper.mledger.impl.ManagedCursorTest
  Error:  org.apache.bookkeeper.mledger.impl.ManagedCursorTest.testForceCursorRecovery  Time elapsed: 0.031 s  <<< FAILURE!
  java.lang.AssertionError: expected [true] but found [false]
  	at org.testng.Assert.fail(Assert.java:110)
  	at org.testng.Assert.failNotEquals(Assert.java:1577)
  	at org.testng.Assert.assertTrue(Assert.java:56)
  	at org.testng.Assert.assertTrue(Assert.java:66)
  	at org.apache.bookkeeper.mledger.impl.ManagedCursorTest.testForceCursorRecovery(ManagedCursorTest.java:4831)
  	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
  	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
  	at org.testng.internal.invokers.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:139)
  	at org.testng.internal.invokers.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:47)
  	at org.testng.internal.invokers.InvokeMethodRunnable.call(InvokeMethodRunnable.java:76)
  	at org.testng.internal.invokers.InvokeMethodRunnable.call(InvokeMethodRunnable.java:11)
  	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
  	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
  	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
  	at java.base/java.lang.Thread.run(Thread.java:1583)
```

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by ZhaoGuorui666**
Fixes #23417 

### Motivation
Ideally, recoverFromLedger should only call operationComplete upon successful recovery and operationFailed upon failure. However, in the actual implementation, both methods may be called due to the following reasons:

Concurrent Execution: If recoverFromLedger is executed in different threads, it may simultaneously meet both success and failure conditions, leading to both callbacks being triggered.

After adding logs, we may see both:
* Cursor recovery success
* Cursor recovery failed

### Modifications

Use an atomic variable. Use an AtomicBoolean in the callback to mark whether the callback method has already been called, preventing duplicate calls.

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete`

### CODE REVIEW

**Comment by nodece** [comment]
Could you verify https://github.com/apache/pulsar/pull/23518/files#diff-42dd67f8328871b3f0d30446ed1a4f1856807860c5226a9ec6c4d3a9f8813428R4882? Hrere should append a `return`.

**Comment by ZhaoGuorui666** [comment]
I understand the `ManagedCursorImpl.recoverFromLedger()` called by the bottom layer of this test, which calls `Bookeeper.syncOpenLedger()` instead of `TestPulsarMockBookKeeper`

image

image

**Comment by ZhaoGuorui666** [comment]
Execute 3000 times, all passed
image

**Review by Technoboy- — APPROVED** [review]

**Comment by nodece** [comment]
In this test, we are using `org.apache.bookkeeper.mledger.impl.ManagedCursorTest.TestPulsarMockBookKeeper`.

`ManagedCursorImpl.recoverFromLedger()` will call this method from that:
```
  public void asyncOpenLedger(final long lId, final DigestType digestType, final byte[] passwd,
                final OpenCallback cb, final Object ctx) {
            if (ledgerErrors.containsKey(lId)) {
                cb.openComplete(ledgerErrors.get(lId), null, ctx);
                return; // TODO: Please add a return.
            }
            super.asyncOpenLedger(lId, digestType, passwd, cb, ctx);
        }
```

It seems that the test results were affected by multiple callback calls.

**Review by nodece — CHANGES_REQUESTED** [review]
Left a comment.

**Comment by ZhaoGuorui666** [comment]
@nodece Thank you for your prompt and suggestion. It has been revised.

**Comment by ZhaoGuorui666** [comment]
> ```
> 2024-10-30T12:38:15,591 - ERROR - [main:ManagedCursorImpl@553] - [my_test_ledger] Error opening metadata ledger -1 for cursor c1: Bookie handle is not available
> 2024-10-30T12:38:15,592 - INFO  - [main:ManagedCursorImpl@750] - [my_test_ledger] Cursor c1 recovered to position 3:-1
> 2024-10-30T12:38:15,592 - INFO  - [main:MetaStoreImpl@257] - expectedVersion:0 [my_test_ledger] Creating consumer c1 on meta-data store with cursorsLedgerId: -1
> markDeleteLedgerId: 3
> markDeleteEntryId: -1
> lastActive: 1730263095578
> 
> 2024-10-30T12:38:15,593 - INFO  - [test-OrderedScheduler-0-0:ManagedCursorImpl@550] - [my_test_ledger] Opened ledger -1 for cursor c1. rc=-7
> 2024-10-30T12:38:15,593 - ERROR - [test-OrderedScheduler-0-0:ManagedCursorImpl@553] - [my_test_ledger] Error opening metadata ledger -1 for cursor c1: No such ledger exists on Bookies
> 2024-10-30T12:38:15,593 - INFO  - [test-OrderedScheduler-0-0:ManagedCursorImpl@750] - [my_test_ledger] Cursor c1 recovered to position 3:-1
> 2024-10-30T12:38:15,593 - INFO  - [test-OrderedScheduler-0-0:MetaStoreImpl@257] - expectedVersion:0 [my_test_ledger] Creating consumer c1 on meta-data store with cursorsLedgerId: -1
> markDeleteLedgerId: 3
> markDeleteEntryId: -1
> lastActive: 1730263095578
> 
> 2024-10-30T12:38:15,601 - INFO  - [bookkeeper-ml-scheduler-OrderedScheduler-1-0:ManagedCursorImpl$25@2772] - ledger.getStore().asyncUpdateCursorInfo operation complete
> 2024-10-30T12:38:15,602 - WARN  - [bookkeeper-ml-scheduler-OrderedScheduler-1-0:ManagedCursorImpl$25@2779] - [my_test_ledger] Failed to update cursor metadata for c1 due to version conflict org.apache.pulsar.metadata.api.MetadataStoreException$BadVersionException:
> 
> java.lang.AssertionError:
> Expected :true
> Actual   :false
> <Click to see difference>
> 
> 
> 	at org.testng.Assert.fail(Assert.java:110)
> 	at org.testng.Assert.failNotEquals(Assert.java:1577)
> 	at org.testng.Assert.assertTrue(Assert.java:56)
> 	at org.testng.Assert.assertTrue(Assert.java:66)
> 	at org.apache.bookkeeper.mledger.impl.ManagedCursorTest.testForceCursorRecovery(ManagedCursorTest.java:4858)
> ```
> 
> From the log of failed test, we can see `expectedVersion:0 [my_test_ledger] Creating consumer c1 on meta-data store with cursorsLedgerId: -1` twice. The root cause of this flaky test is we **call the cb.openComplete() twice**, which may cause race condition: when the first cursorInfo has not been saved, the second update cursorInfo started use the oldVersion, so when to save the second cursorInfo meta, it will throw BadVersion exception
> 
> So We don't need to add `AtomicBoolean callbackInvoked = new AtomicBoolean(false);`, we can just fix this to add return in TestPulsarMockBookKeeper.asyncOpenLedger or add else like
> 
> ```java
> public void asyncOpenLedger(final long lId, final DigestType digestType, final byte[] passwd,
>         final OpenCallback cb, final Object ctx) {
>     if (ledgerErrors.containsKey(lId)) {
>         cb.openComplete(ledgerErrors.get(lId), null, ctx);
>     } else {
>         super.asyncOpenLedger(lId, digestType, passwd, cb, ctx);
>     }
> }
> ```

Yes, doing an if else or adding a return is the way to truly solve the problem. Here are the test results:
﻿
企业微信截图_17302657085370

**Review by nodece — APPROVED** [review]
LGTM

File: managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/ManagedCursorTest.java [hunk 1]
```diff
@@ -4874,8 +4873,9 @@ public void asyncOpenLedger(final long lId, final DigestType digestType, final b
                 final OpenCallback cb, final Object ctx) {
             if (ledgerErrors.containsKey(lId)) {
                 cb.openComplete(ledgerErrors.get(lId), null, ctx);
+            }else{
```
Comment (lhotari) [inline managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/ManagedCursorTest.java:4876]: please use the default formatting code style that we use in Pulsar code base, [instructions for setting it up in IntelliJ](https://pulsar.apache.org/contribute/setup-ide/#configure-code-style).

Comment (nodece) [inline managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/ManagedCursorTest.java:4876]:
```suggestion
            } else {
```

Fix code style.

**Review by lhotari — APPROVED** [review]
LGTM, good work @ZhaoGuorui666. Thanks for putting effort in fixing flaky tests.

**Pull request merged**